### PR TITLE
feat: custom d3 number locale

### PR DESCRIPTION
**Pull request by ebaratte**
### SUMMARY

This PR is an attempt to provide a way of customising the format of the numbers and currency displayed in superset frontend.
It is linked to (at least) issues #3972, #18938, #11328, #8913
Please also look at PR #17796, which provides a different approach to the same issue.  

Design choice:
- the user can change the following d3 **number** formatting parameters: decimal, thousands, grouping, currency, numerals, percent, minus and nan
- this change does not include the **time formatting** parameters, since I believe it to be a different issue
- unlike #17796, it does not tie the format to the `LANGUAGES` config key: the user can use different parameters for the frontend language and number formatting

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Before
image

After:
image

### TESTING INSTRUCTIONS
- in config.py, add a D3_FORMAT dictionary with [d3 locale settings](https://github.com/d3/d3-format/blob/main/README.md#locale_format) 
- check the number formatting on any chart

### ADDITIONAL INFORMATION
- [x] Has associated issue: #3972, #18938, #11328, #8913
- [x] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by k7daniel** [comment]
This is very cool. Would be great to just set D3_FORMAT and pass that in to the docker image.

**Comment by SergeRomankevich** [comment]
Tables with different currencies are crying aside and waiting for the merge

**Comment by enryls** [comment]
Great feature! Hope to see it merged soon

**Comment by pm-at** [comment]
Is there a way to speed up the review process?

**Comment by theYaro** [comment]
Waiting for this merge.

**Comment by pm-at** [comment]
I'm sorry for repeated nudge, but this is a really helpful feature. Waiting for this PR to be merged..

**Comment by lf-floriandin** [comment]
Would also love if this get reviewed and merged!

**Comment by villebro** [comment]
@ebaratte would you be able to rebase this PR to fix the conflicts?

**Comment by Pedrolrbr** [comment]
@justinpark @eschutho @kgabryje @villebro any updates?

File: superset/config.py
Comment (villebro) [inline superset/config.py:390]: nit: as `TypedDict` is now a default thing in our currently supported Python versions, I'd prefer to create a `D3Format(TypedDict)` for this.

File: superset-frontend/src/types/bootstrapTypes.ts
Comment (villebro) [inline superset-frontend/src/types/bootstrapTypes.ts:154]: Do we really need `Partial` here? It appears that the added `dict` in `config.py` defines all properties of `FormatLocaleDefinition`. If not, let's make sure the type in `config.py` reflects the partialness.

**Review by villebro — COMMENTED** [review]
I know that Superset is badly lacking in this space, but the main issue I have with this approach is the fact that it assumes the deployment as a whole speaks one single locale. As we start to prepare for making Superset more localizable (see e.g. a WIP PR of mine that adds timezone support to the chart data API: #23511), it would be nice if we could start making gradual progress towards having user or chart specific locales. Let's say we want to have one chart that shows euros, and another that shows US dollars. This would not be possible with this setting.

Having said that, I think this approach is good to cover use cases where the entire deployment wants to "speak" only one specific locale, as long as we leave open the door for making this overridable per user/chart. And based on my review this doesn't paint us into any corner, so I think this looks good. Thoughts @ebaratte @Pedrolrbr @kgabryje @rusackas ?

**Comment by ebaratte** [comment]
> I know that Superset is badly lacking in this space, but the main issue I have with this approach is the fact that it assumes the deployment as a whole speaks one single locale. As we start to prepare for making Superset more localizable (see e.g. a WIP PR of mine that adds timezone support to the chart data API: #23511), it would be nice if we could start making gradual progress towards having user or chart specific locales. Let's say we want to have one chart that shows euros, and another that shows US dollars. This would not be possible with this setting.
> 
> Having said that, I think this approach is good to cover use cases where the entire deployment wants to "speak" only one specific locale, as long as we leave open the door for making this overridable per user/chart. And based on my review this doesn't paint us into any corner, so I think this looks good. Thoughts @ebaratte @Pedrolrbr @kgabryje @rusackas ?

This is indeed the general idea of this PR: set default values for the whole instance. This is especially relevant for the currency, since there is no proper way at the moment to change the default currency, and displaying an incorrect currency makes the charts incorrect - whereas number formatting is more of a cosmetic issue. Obviously, it assumes that all the data is in a single currency. 

It shouldn't prevent local overrides, like:
- setting a specific currency on a chart (#22374)
- using a number format that depends on the current locale (#17796), or via user settings

**Comment by villebro** [comment]
> This is indeed the general idea of this PR: set default values for the whole instance. This is especially relevant for the currency, since there is no proper way at the moment to change the default currency, and displaying an incorrect currency makes the charts incorrect - whereas number formatting is more of a cosmetic issue. Obviously, it assumes that all the data is in a single currency.
> 
> It shouldn't prevent local overrides, like:
> 
> * setting a specific currency on a chart ([feat(formatters): [WIP] Adding a few currency formatters #22374](https://github.com/apache/superset/pull/22374))
> * using a number format that depends on the current locale ([feat: Create D3NumberFormatters with active Locale #17796](https://github.com/apache/superset/pull/17796)), or via user settings

Ok you don't need to convince me more @ebaratte 🙂 Looks good to go to me, but there appears to be some frontend UTs that are failing. Please take a look so we can get this in 👍

**Review by villebro — APPROVED** [review]
LGTM once the tests are passing.

**Comment by k7daniel** [comment]
Yay! Can't wait to test this out!!

**Comment by rusackas** [comment]
Hoping to test this on an ephemeral build, but then I can't wait to click the Merge button on this!!!! 🚀

**Comment by rusackas** [comment]
Sorry for all the noise on the thread, folks! I'm trying to troubleshoot the ephemeral environment builds, so it needs a little kicking ;)

**Comment by rusackas** [comment]
Had a nice long chat with @mistercrunch about this approach and a couple other approaches to improving currency handling in Superset. At the end of the day, I think all three have their place. Expect some more proposals and/or PRs in the future. In the meantime, this looks ready to go. Thanks for the PR!!!! Hitting the merge button!

**Comment by cwegener** [comment]
Just cross-linking the alternative designs for future reference (somehow Github doesn't place links from Discussion threads for issue mentions into the issue log automatically ...)
https://github.com/apache/superset/discussions/23684

**Comment by phifa** [comment]
So I am not familiar with `D3_FORMAT`. Can someone tell me what I need to add to `config.py`? I want this number to be displayed in a european format, like this: 579 781 or like this 579.781,000. Thanks a ton!

**Comment by cwegener** [comment]
@phifa 
According to the documentation, this config should be what is needed:
```
D3_FORMAT = {
    "decimal": ".",
    "thousands": ",",
    "grouping": [3],
    "currency": ["€", ""]
}
```

**Comment by cwegener** [comment]
> Does it work on version 2.1.0 ? Nothing changes for me.

No. This PR is not part of 2.1

This PR is only available in the master branch.

**Pull request merged**
